File: 127. Word Ladder/solution.java
class Solution {
    class Node {
        String string;
        int length;
        public Node(String string, int length) {
            this.string = string;
            this.length = length;
        }
    }
    
    public int ladderLength(String beginWord, String endWord, List<String> wordList)    {
        Set<String> set = new HashSet<>(wordList);
        int result = Integer.MAX_VALUE;
        Queue<Node> q = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        q.offer(new Node(beginWord, 1));
        visited.add(beginWord);
        while (!q.isEmpty()) {
            Node node = q.poll();
            // System.out.println(node.string);
            if (node.string.equals(endWord)) {
                result = Math.min(result, node.length);
                continue;
            }
            
            String word = node.string;
            for (int i = 0; i < word.length(); i++) {
                StringBuilder newString = new StringBuilder(word);
                for (int j = 0; j < 26; j++) {
                    newString.setCharAt(i, (char) ('a' + j));
                    String newWord = newString.toString();
                    if (!visited.contains(newWord) && set.contains(newWord)) {
                        q.offer(new Node(newWord, node.length + 1));
                        visited.add(newWord);
                    }
                }
            }
        }
        return result == Integer.MAX_VALUE ? 0 : result;
    }
}

File: 127. Word Ladder/solution.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        q = [(beginWord, 1, [beginWord])]
        # ret = math.inf
        visited = set()
        visited.add(beginWord)
        dic = set(wordList)
        while len(q):
            w = q.pop(0)
            word = w[0]
            # print(word)
            length = w[1]
            if word == endWord:
                # print(w[2])
                # ret = min(ret, length)
                return length
            else:
                word = list(word)
                # print(word)
                for i in range(len(word)):
                    # print(i)
                    old_char = word[i]
                    for c in 'abcdefghijklmnopqrstuvwxyz':
                        word[i] = c
                        new_word = ''.join(word)
                        # print(new_word)
                        if new_word not in visited and new_word in dic:
                            new_paths = w[2].copy()
                            new_paths.append(new_word)
                            q.append((new_word, length + 1, new_paths))
                            visited.add(new_word)
                    word[i] = old_char
        return 0
